fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationsDialogStore.kt

```diff
@@ -0,0 +1,219 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+package org.mozilla.fenix.translations
+
+import mozilla.components.concept.engine.translate.Language
+import mozilla.components.concept.engine.translate.TranslationError
+import mozilla.components.lib.state.Action
+import mozilla.components.lib.state.Middleware
+import mozilla.components.lib.state.State
+import mozilla.components.lib.state.Store
+
+/**
+ * The [TranslationsDialogStore] holds the [TranslationsDialogState] (state tree).
+ *
+ * The only way to change the [TranslationsDialogState] inside
+ * [TranslationsDialogStore] is to dispatch an [Action] on it.
+ */
+class TranslationsDialogStore(
+    initialState: TranslationsDialogState,
+    middlewares: List<Middleware<TranslationsDialogState, TranslationsDialogAction>> = emptyList(),
+) : Store<TranslationsDialogState, TranslationsDialogAction>(
+    initialState,
+    ::translationsDialogStateReducer,
+    middlewares,
+) {
+    init {
+        dispatch(TranslationsDialogAction.ExpectedTranslations)
+    }
+}
+
+/**
+ * The current state of the Translations bottom sheet dialog.
+ *
+ * @property isTranslated The page is currently translated.
+ * @property isInProgress The page is currently attempting a translation.
+ * @property positiveButtonType Can be enabled,disabled or in progress.
+ * @property error An error the can occur at translation process.
+ * @property dismissDialogState Whether the dialog bottom sheet should be dismissed.
+ * @property showFirstTime Whether translations first flow should be shown.
+ * @property initialFrom Initial "from" language, based on the translation state and page state.
+ * @property initialTo Initial "to" language, based on the translation state and page state.
+ * @property fromLanguages Translation menu items to be shown in the translate from dropdown.
+ * @property toLanguages Translation menu items to be shown in the translate to dropdown.
+ */
+data class TranslationsDialogState(
+    val isTranslated: Boolean = false,
+    val isInProgress: Boolean = false,
+    val positiveButtonType: PositiveButtonType? = null,
+    val error: TranslationError? = null,
+    val dismissDialogState: DismissDialogState? = null,
+    val showFirstTime: Boolean = false,
+    val initialFrom: Language? = null,
+    val initialTo: Language? = null,
+    val fromLanguages: List<Language>? = null,
+    val toLanguages: List<Language>? = null,
+) : State
```

Comment:
I would suggest considering whether this can be wrapped into the existing `TranslationEngineState` that exists in the browser store, and filtering in the `BrowserFragment` based on the currently selected tab.

Reply:
Thanks for taking a look at the architecture on this patch! 
>wrapped into the existing TranslationEngineState 

Could you give an example or psudocode to illustrate?

I don't have a sense of what the architecture should be here, but have some relevant details to guide the discussion:

* `TranslationEngineState`'s lifecycle is tightly coupled to the browser translations engine and a [delegate](https://searchfox.org/mozilla-mobile/rev/4d9cf33f0069fd4650a7598e6c07dcfcd86c86d1/firefox-android/android-components/components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/translate/GeckoTranslateSessionDelegate.kt#55) fully controls the lifecycle of those data properties. I'd say `TranslationEngineState` and `TranslationsState` as they stand are not implementation specific. [TranslationsState](https://searchfox.org/mozilla-mobile/source/firefox-android/android-components/components/browser/state/src/main/java/mozilla/components/browser/state/state/TranslationsState.kt) is less coupled to the translations engine itself. [TranslationEngineState](https://searchfox.org/mozilla-mobile/source/firefox-android/android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/translate/TranslationEngineState.kt) is tightly coupled to the translations engine.

* My impression of `TranslationsDialogState` is that it is an implementation specific repackaging of the data model. It is pulling the raw data from the BrowserStore and preparing it for the Fenix implementation. For example:

    `TranslationsDialogState`:
    `isTranslated` -> `TranslationsState.isTranslated`
    `isInProgress` -> `TranslationsState.isTranslateProcessing`
    `positiveButtonType` -> No engine correlation
    `error` -> `TranslationsState.translationError`
    `dismissDialogState` -> No engine correlation
    `showFirstTime` -> Fenix controlled concept
    `initialFrom`-> `TranslationsState.translationEngineState.initialFromLanguage`
    `initialTo` ->`TranslationsState.translationEngineState.initialToLanguage`
    `fromLanguages` -> `TranslationsState.supportedLanguages.fromLanguages`
    `toLanguages` -> `TranslationsState.supportedLanguages.toLanguages`
    

* I'm not sure what I make of the two sources of truth for the state of things like `isTranslated` especially, that keeps catching my eye as something to examine.

Reply:
Ah, thanks for clarifying about `TranslationEngineState` vs `TranslationState`. 

> My impression of TranslationsDialogState is that it is an implementation specific repackaging of the data model.

Do we need to repackage it? There's no reason that all the properties that are replicated across the `TranslationsDialogState` and `TranslationState` need to be. We can observe them directly from the `BrowserStore`, which would be my suggestion so that we aren't doing extra work in the observations and re-dispatching of the state. That avoids the risk of keeping two sources of truth in sync, as you've mentioned.

As to the other properties in `TranslationDialogState` that are Fenix-specific:

Some of these are pretty simple derived properties of the state that already exists in `BrowserStore`. For example, `positiveButtonType` could become a `when` expression:
```
// TranslationDialogFragment
val translationState = browserStore.observeAsComposableState { state.findTab(<currentTabId>).translationState }
val positiveButtonType = when {
    translationsState.isTranslated && !translationsState.isTranslateProcessing -> PositiveButtonType.Enabled
    is TranslationError -> PositiveButtonType.Disabled
}
```

This isn't the exact logic of course, but just an example. We don't necessarily need to keep derived properties in a separate state class if they are just used to display different UI conditionally. I think a similar case could be made for `dismissDialogState`. `showFirstTime` is something that could be passed directly from the Fragment to the Composable IMO. 

The benefit to having a `TranslationDialogStore` is that we can actually unit test some of the derived state logic. However, I am not seeing unit tests for that here, so we should either do the simple thing and remove the additional Store or do the "correct" thing and write tests as well.

Reply:
> However, I am not seeing unit tests for that here, so we should either do the simple thing and remove the additional Store or do the "correct" thing and write tests as well

Reflecting on this comment from yesterday, I didn't mean for this part to read facetious or anything, hope it didn't!

I do think having a `TranslationDialogStore` with some of these derived states in them will be useful long-term and lets us have better tests, so is the right thing to do. That said, I am still of the opinion that we don't need to replicate the state that's already hosted in `BrowserStore`.

Reply:
@iorgamgabriel It looks like this patch went the direction of adding tests. Could you also add a comment explaining the reasoning behind that direction too? 

Based on the conversation we had in Slack, the reasoning you mentioned there made sense to me; however, I'm not the best contact to be reviewing Fenix architecture. Please post some additional information here as well for other reviewers with much more Fenix experience!

I'm more comfortable with the duplicated store portions based on having tests. If you can think of a way to extract `isTranslated` and `isInProgress` states, I'd be more comfortable, but it sounds like that might not be the best direction based on your research.

Reply:
I added unit tests for TranslationsDialogBinding, TranslationsDialogMiddleware and TranslationsDialogReducer.
I tried to remove almost all the business logic from the fragment. 
I removed the part when the translation icon from the toolbar is violet at the end of the translation task (I will do it in other pr).